### DISCUSSION TITLE
Kids friendly option - use A1111 plugin to filter out nsfw content?

### DISCUSSION THREAD

**Discussion by VictorZakharov** [Kilvoctu/aiyabot#108] [General]
I installed the plugin in A1111 and it works, all nsfw content is blacked out.
Is it possible to make Aiyabot use it also?

EDIT: I tried adding built-in nsfw filter in discord, it filters my messages, but doesn't filter those coming from Aiya.

**Comment by Kilvoctu** [comment]
Heya, generally I don't support any extensions to Web UI (anything in [this list)](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Extensions#extensions-category) as they're out of scope. However, NSFW checker is sort of a basic thing, so I looked into it.  
> Is it possible to make Aiyabot use it also?  

As far as I can tell, it's not possible. There are no API endpoints and nothing in Web UI config for AIYA to connect to.  
It also doesn't appear in the Web UI script list, so I can't add something like `"script_name": "NSFW check"` to the payload.  
Automatic1111 moved the feature out of repo into an [extension](https://github.com/AUTOMATIC1111/stable-diffusion-webui-nsfw-censor). If you're able to convince them to add API support for it, I'll try to implement it.

In the meantime, I pushed a feature with commit 64dac4cc02cc06648a9360fea1e273f0b6c6ace9. When you launch AIYA, she'll add `prompt_ban_list = []` to the config.toml.
You can fill it out with any words you don't want to have processed, then restart bot. Afterward, she won't generate an image if any of the words are found in prompt.   

For example, I can enter the value like so in config.toml: `prompt_ban_list = ["unsavory thing"]`
And if I try to include it in prompt or 🖋️...
![image](https://user-images.githubusercontent.com/2993060/218341493-f17d4d03-5da1-482d-b0bc-b0256b60170c.png)
Any matches of the string will trigger, so if you ban `star` it'll catch `start` too.

It's not perfect, but it should be better than nothing.

**Comment by VictorZakharov** [comment]
@Kilvoctu I see. Prompt blacklist is good, but doesn't completely solve the problem. With most good models, asking for a female, full body will often result in nsfw content, even without explicitly asking for it. Similarly when asking for a female, outstanding beauty, pretty etc. And even when drawing character female action shots (RPG, movie style etc), the content will not necessarily be nsfw, but very suggestive, so A1111's nsfw checker will block it also, which I think is the right thing to do.

Why can I not use discord's nsfw blocker? It blocks my posts, but not the bot?

**Comment by Kilvoctu** [comment]
Yeah, I'm aware it's not perfect. It's the best I can do with what I have available. Not sure what more I can do; I only manage this repo. 🤷‍♀️

> so A1111's nsfw checker will block it also, which I think is the right thing to do.

Yes, it's ideal, but you need to contact A1111 or leave feedback on that [extension repo](https://github.com/AUTOMATIC1111/stable-diffusion-webui-nsfw-censor). AIYA is not integrated into Web UI, like the stuff in the extensions list. She's a separate application that uses the API or can read through Web UI's config. 

> Why can I not use discord's nsfw blocker? It blocks my posts, but not the bot?  

This is a Discord feature, so you'll want to ask Discord Support. I simply don't know.

**Comment by VictorZakharov** [comment]
I created a ticket with A1111:
- https://github.com/AUTOMATIC1111/stable-diffusion-webui-nsfw-censor/issues/1

Not sure if it's the right place, because I was assigned ticket number 1, meaning no other tickets were created before this.

I also messaged discord support and asked them about this issue. I will keep this thread updated if I find anything. Thanks!

**Comment by Kilvoctu** [comment]
Moving this over to Discussions, since I'd prefer Issues to be for bug reports

**Comment by VictorZakharov** [comment]
@Kilvoctu Can you add the following configs?
- prompt_ignore_list - will remove these words from the prompt and send the rest, not telling the user, or telling the user, either is fine.
- ~prompt_prefix - will add a fixed prefix, in my case sfw, but it has other uses.~ (see below EDIT)
- prompt_neg_prefix- will add a prefix to the negative prompt, in my case nsfw.

I think this should help silently make the pictures safer, without blocking/stopping the user.
Some cloud SDs are already doing this, auto-converting nsfw prompt to sfw automatically.

What do you think?

EDIT: did some local experiments with various models, not all of them are horny by default, but some are. I don't think `prompt_prefix` (sfw) is needed for my use case, but `prompt_neg_prefix` (nsfw) can solve it. However, for `prompt_neg_prefix` to work, `prompt_ignore_list` should work also.

It seems that the positive prompt takes precedence over negative prompt, so for any conflicts the positive prompt is used, which means someone can essentially override negative prompt if they are persistent enough. So removing trigger words from the negative prompt + adding negative prefix seems to address the problem in its entirety.

BTW, haven't heard from Discord on this issue.

**Reply by Kilvoctu** [reply]
Heya, I added both `prompt_ignore_list` and `prompt_neg_prefix` with 86b0d9dcb7251a82a65122677597d748df2d5fd5

`prompt_ignore_list` will do as you described. It'll look through and match any words in the list and remove from prompt, then proceed normally. Conveniently AIYA has always had ability to hide the true prompt from user (because she hides activator tokens). So user will not know their prompt had words removed unless they check the metadata.

`prompt_neg_prefix` adds the configured words to the beginning of any negative prompt. The user cannot override this.

I tested it quite a bit and should be working as expected, but if there's any bugs or unintended behaviors you can follow up.

**Reply by VictorZakharov** [reply]
@Kilvoctu This worked great, thank you! Can you hide the negative prompt in bot's responses in discord if all it contains is `negative_prompt_prefix`? Would be great to have it as transparent as possible.

Also now that I'm thinking about it, is it possible to remove unsafe words from bot's discord reply/confirmation also? Both negative and positive prompts. Could be another option or default.

**Reply by Kilvoctu** [reply]
> is it possible to remove unsafe words from bot's discord reply/confirmation also? Both negative and positive prompts. Could be another option or default.

Thought you said either was fine for telling the user, lol 🤔
Think I prefer to obfuscate what words were removed from reply/confirmation.  
To implement this request, I'd have to rewrite significant parts of the code for this one use case, which isn't something I want to do; or I change how prompts are displayed for every bot host, which is not my desired behavior for the aiyabot.

If you'd like to add this, you can do the latter and somewhere on these lines add `simple_prompt = mod_results[1]` https://github.com/Kilvoctu/aiyabot/blob/86b0d9dcb7251a82a65122677597d748df2d5fd5/core/stablecog.py#L231-L232

`prompt` is what's sent to the Web UI. The end user never sees this string.
`simple_prompt` is the string that's actually displayed to the end user instead. By putting `simple_prompt` there, it'll run it through the mod function and remove the words like it does for `prompt`; the mod changes also would be reflected on subsequent generations with 🖋️ and 🎲 (and shown on📋).

---

> Can you hide the negative prompt in bot's responses in discord if all it contains is `negative_prompt_prefix`? Would be great to have it as transparent as possible.

Wouldn't that make it less transparent to the end user what was altered??  
Anyway, how AIYA displays negative prompts was already bothering me a bit, so I was thinking about adjusting how she displays it. A lot of servers set large paragraphs into the per-channel defaults for negative prompt, and I think it's unsightly to see on each generation. I'll see if I can resolve that and this in one go later on. I put it on my to do list (#85)

**Reply by VictorZakharov** [reply]
@Kilvoctu That's what I originally wanted, yes. But then I tried it and figured that it might be best to show the prompt as it was drawn, rather than how it was typed.

Let's imagine the situation, we have older kids and younger kids using stable diffusion with the bot. Older kids know some words than younger kids don't. Do I want younger kids to learn those words? Not really. I'd rather show a clean prompt so that everyone can learn together as if the typed prompt was safe in the first place. Also in this case I don't need to moderate the channel. Well, not for image generation anyway.

Regarding negative prompt, it's unsightly, yes, even if users entered something there. Also my experience with kids is that they don't really use it, so it just adds confusion. Which is why I liked the bot idea, fewer options = easy to learn. Rather than throwing thousands of dialog options of A1111 web UI at them. There is only /draw, to start.

Your solution for modifying stablecog.py worked great. I see some leading and trailing empty spaces which should probably be removed, but it's fine, the world will survive.

I also wrote my first python line today, which makes me officially a python developer. Thanks for that! Feeling adventurous, I also commented out lines 264 and 265, which effectively removes the negative prompt from bot reply/confirmation.

```python
# if negative_prompt != '':
# reply_adds += f'\nNegative Prompt: ``{negative_prompt}``'
```

**Reply by Kilvoctu** [reply]
>  I also commented out lines 264 and 265, which effectively removes the negative prompt from bot reply/confirmation.

Yup, that was the line I was going to edit, too, with some conditions to see if nprompt was from user or pre-filled.

The main goal of AIYA was being easy for end user (_as a result, a lot of the tedious stuff is placed upon the bot host_), both in operation as well as what's displayed. It's originally why the reply/confirmation has so many conditions as you noticed, and I haven't revised it since the initial implementation.  

> it might be best to show the prompt as it was drawn, rather than how it was typed.  

I might provide a proper config option for this later when handling the reply/confirmation thing for negative prompt.  

After providing you that 1 line edit, I was thinking about my code and it might be simpler than previously expected, not sure 🤔. I was actually a bit surprised that 1 line had such a profound affect on everything that followed it. Can try to add in some conditionals and clean up the output.

**Reply by VictorZakharov** [reply]
@Kilvoctu The power of an open source community! Glad I could help some. BTW, I revised my edit near line 233 to this:

```python
import re # at the top of stablecog.py
#... other code
#... line 233
simple_prompt = mod_results[1].strip()  # remove leading and trailing spaces
simple_prompt = re.sub(r'\s+', ' ', simple_prompt)  # replace multiple spaces with single space
```

I generated the above code with ChatGPT. Seems to work, didn't do extensive testing.

**EDIT:** My changes are available here in case anyone wants to use them.
- https://github.com/VictorZakharov/aiyabot/tree/victor/tweaks

Or more specifically:
- https://github.com/Kilvoctu/aiyabot/commit/65ac0e54c8932fd1936ff604f364a72677fb7210

**Reply by Kilvoctu** [reply]
Heya I added everything mentioned before with PR #112.

It adds `display_ignored_words` to config, so host can choose if they want to show the ignored words. I set it to `"False"` by default so you don't need to do anything extra.  
Initial reply/confirmation will also be a lot cleaner. It won't show parameters that are pre-filed by AIYA. The words in `negative_prompt_prefix` aren't displayed either.

The simple edits we did weren't working properly if user is using 🖋️, so most of the code action happens in the prompt_mod function. To handle spacing, I used something aside from RegEx.

Thanks for your input 👍👍

--- 

~~The 🖋️ and 📋 buttons still shows `negative_prompt_prefix` words... I'll revise it later.~~
Done 2e197d89ae7f966dfbfe19b634fde65ba88802f6

**Reply by VictorZakharov** [reply]
@Kilvoctu Works great, thanks! I'm now using the main branch instead of my custom branch.
